zilliqa/src/contracts/deposit_v3.sol

```diff
@@ -0,0 +1,685 @@
+                        .stakerKeys[j];
+                    $._committee[i % 3].stakers[
+                        stakerKey
+                    ] = latestComputedCommittee.stakers[stakerKey];
+                }
+            }
+
+            $.latestComputedEpoch = currentEpoch() + 2;
+        }
+    }
+
+    // Returns the next block number at which new stakers are added,
+    // existing ones removed and/or deposits of existing stakers change
+    function nextUpdate() public view returns (uint256 blockNumber) {
+        DepositStorage storage $ = _getDepositStorage();
+        if ($.latestComputedEpoch > currentEpoch())
+            blockNumber = $.latestComputedEpoch * $.blocksPerEpoch;
+    }
+
+    // keep in-sync with zilliqa/src/precompiles.rs
+    function _blsVerify(
+        bytes memory message,
+        bytes memory pubkey,
+        bytes memory signature
+    ) internal view returns (bool) {
+        bytes memory input = abi.encodeWithSelector(
+            hex"a65ebb25", // bytes4(keccak256("blsVerify(bytes,bytes,bytes)"))
+            message,
+            signature,
+            pubkey
+        );
+        uint256 inputLength = input.length;
+        bytes memory output = new bytes(32);
+        bool success;
+        assembly {
+            success := staticcall(
+                gas(),
+                0x5a494c81, // "ZIL\x81"
+                add(input, 0x20),
+                inputLength,
+                add(output, 0x20),
+                32
+            )
+        }
+        require(success, "blsVerify");
+        bool result = abi.decode(output, (bool));
+        return result;
+    }
+
+    function deposit(
+        bytes calldata blsPubKey,
+        bytes calldata peerId,
+        bytes calldata signature,
+        address rewardAddress
+    ) public payable {
+        if (blsPubKey.length != 48) {
+            revert UnexpectedArgumentLength("bls public key", 48);
+        }
+        if (peerId.length != 38) {
+            revert UnexpectedArgumentLength("peer id", 38);
+        }
+        if (signature.length != 96) {
+            revert UnexpectedArgumentLength("signature", 96);
+        }
+        DepositStorage storage $ = _getDepositStorage();
+
+        uint64 chain_id = uint64(block.chainid);
+        bytes memory message = new bytes(blsPubKey.length + 8 + 20); // blsPubKey + uint64 + address
+
+        for (uint256 i = 0; i < blsPubKey.length; i++) {
+            message[i] = blsPubKey[i];
+        }
+
+        for (uint256 i = 0; i < 8; i++) {
+            message[blsPubKey.length + i] = bytes1(
+                uint8(chain_id >> (8 * (7 - i)))
+            );
+        }
+
+        for (uint256 i = 0; i < 20; i++) {
+            message[blsPubKey.length + 8 + i] = bytes20(msg.sender)[i];
+        }
+
+        // Verify bls signature
+        if (!_blsVerify(message, blsPubKey, signature)) {
+            revert RogueKeyCheckFailed();
+        }
+
+        Staker storage staker = $._stakersMap[blsPubKey];
```

Comment:
nit: does it make sense to move this a few lines down, where `staker` is first used?